File: java/src/nuTetris/PieceFactory.java
/*
 *  This file is part of nuTetris
 *  Author: <antonino.calderone@gmail.com>
 *
 */

package nuTetris;

import java.util.Vector;

/**
 * Holds pieces catalog and provides a copy of each of them based
 * on a given index number
 */
public class PieceFactory {

    /** Return the factory instance */
    public static PieceFactory getInstance() {
        return instance;
    }

    /** Clone a new piece */
    Piece make(int pieceIdx) {
        return catalog.elementAt(pieceIdx).clone();
    }

    /** Get the count of pieces within the catalog */
    public int getPieceCatalogSize() {
        return catalog.size();
    }

    /** Add a piece into the catalog */
    protected void addCatalog(Piece piece) {
        catalog.addElement(piece);
    }

    /** Construct a catalog of pieces */
    protected PieceFactory() {
        int i = 0;

        ConcretePieceL L = new ConcretePieceL(++i);
        addCatalog(L);

        ConcretePieceI I = new ConcretePieceI(++i);
        addCatalog(I);

        ConcretePieceJ J = new ConcretePieceJ(++i);
        addCatalog(J);

        ConcretePieceO O = new ConcretePieceO(++i);
        addCatalog(O);

        ConcretePieceS S = new ConcretePieceS(++i);
        addCatalog(S);

        ConcretePieceZ Z = new ConcretePieceZ(++i);
        addCatalog(Z);

        ConcretePieceT T = new ConcretePieceT(++i);
        addCatalog(T);
    }

    private Vector<Piece> catalog = new Vector<Piece>();
    private static PieceFactory instance = new PieceFactory();
}

File: csharp/nuTetris/PieceFactory.cs
using System.Collections.Generic;
using System.Linq;

namespace nuTetris
{
    /**
     * Holds pieces catalog and provides a copy of each of them based
     * on a given index number
     */
    public class PieceFactory
    {
        /** Return the factory instance */
        public static PieceFactory GetInstance() => instance;

        /** Clone a new piece */
        public Piece Make(int pieceIdx) => catalog.ElementAt(pieceIdx).Clone();

        /** Get the count of pieces within the catalog */
        public int PieceCatalogSize => catalog.Count();

        /** Add a piece into the catalog */
        protected void AddCatalog(Piece piece) => catalog.Add(piece);

        /** Construct a catalog of pieces */
        public PieceFactory()
        {
            AddCatalog(new ConcretePieceL(1));
            AddCatalog(new ConcretePieceI(2));
            AddCatalog(new ConcretePieceJ(3));
            AddCatalog(new ConcretePieceO(4));
            AddCatalog(new ConcretePieceS(5));
            AddCatalog(new ConcretePieceZ(6));
            AddCatalog(new ConcretePieceT(7));
        }

        private List<Piece> catalog = new List<Piece>();
        private static PieceFactory instance = new PieceFactory();
    }
}
